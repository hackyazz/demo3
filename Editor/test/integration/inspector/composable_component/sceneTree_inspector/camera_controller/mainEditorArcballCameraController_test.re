open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

open MainEditorTransform;

let _ =
  describe("MainEditor ArcballCameraController", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();

      MainEditorSceneTool.initState(~sandbox, ());
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test set currentSceneTreeNode to be camera", () => {
      beforeEach(() => {
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setSceneFirstCameraToBeCurrentSceneTreeNode,
        );

        CurrentSelectSourceEditorService.setCurrentSelectSource(
          SceneTreeWidgetService.getWidget(),
        )
        |> StateLogicService.getAndSetEditorState;
      });

      test("test ui", () => {
        MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();

        BuildComponentTool.buildInspectorComponent(
          TestTool.buildEmptyAppState(),
          InspectorTool.buildFakeAllShowComponentConfig(),
        )
        |> ReactTestTool.createSnapshotAndMatch;
      });

      describe("test change arcballCameraController distance", () => {
        test("test change should set into engine", () => {
          MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
          let currentGameObjectArcballCamera =
            GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
          let value = 21.1;

          MainEditorArcballCameraControllerTool.changeDistanceAndBlur(
            ~cameraController=currentGameObjectArcballCamera,
            ~value,
            (),
          );

          ArcballCameraEngineService.unsafeGetArcballCameraControllerDistance(
            currentGameObjectArcballCamera,
          )
          |> StateLogicService.getEngineStateToGetData
          |. FloatService.truncateFloatValue(5)
          |> expect == value;
        });

        describe("if blur", () =>
          describe("not refresh inspector", () => {
            let _prepareAndExec = dispatchFuncStub => {
              MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
              MainEditorTransformTool.setLocalEulerAngleData();
              let currentGameObjectArcballCamera =
                GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
              let value = 21.1;

              let dispatchFuncCallCountBeforeChangeTarget =
                dispatchFuncStub |> getCallCount;
              MainEditorArcballCameraControllerTool.changeDistanceAndBlur(
                ~cameraController=currentGameObjectArcballCamera,
                ~value,
                (),
              );

              dispatchFuncCallCountBeforeChangeTarget;
            };

            test(
              "shouldn't remove current scene tree node->local euler angle data",
              () => {
              _prepareAndExec(ReactTool.createDispatchFuncStub(sandbox));

              MainEditorTransformTool.judgeShouldRemoveLocalEulerAngleData()
              |> expect == false;
            });
            test("not refresh inspector", () => {
              let dispatchFuncStub =
                ReactTool.createDispatchFuncStub(sandbox);

              let dispatchFuncCallCountBeforeChangeTarget =
                _prepareAndExec(dispatchFuncStub);

              dispatchFuncStub
              |> withOneArg(
                   AppStore.UpdateAction(Update([|UpdateStore.Inspector|])),
                 )
              |> getCallCount
              |> expect == dispatchFuncCallCountBeforeChangeTarget;
            });
          })
        );

        describe("if drag drop", () =>
          describe("refresh inspector", () => {
            let _prepareAndExec = dispatchFuncStub => {
              MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
              MainEditorTransformTool.setLocalEulerAngleData();
              let currentGameObjectArcballCamera =
                GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
              let value = 21.1;

              let dispatchFuncCallCountBeforeChangeTarget =
                dispatchFuncStub |> getCallCount;
              MainEditorArcballCameraControllerTool.changeDistanceAndDragDrop(
                ~cameraController=currentGameObjectArcballCamera,
                ~changeValue=value,
                ~dragDropValue=value,
                (),
              );

              dispatchFuncCallCountBeforeChangeTarget;
            };

            test(
              "should remove current scene tree node->local euler angle data",
              () => {
              _prepareAndExec(ReactTool.createDispatchFuncStub(sandbox));

              MainEditorTransformTool.judgeShouldRemoveLocalEulerAngleData()
              |> expect == true;
            });
            test("refresh inspector", () => {
              let dispatchFuncStub =
                ReactTool.createDispatchFuncStub(sandbox);

              let dispatchFuncCallCountBeforeChangeTarget =
                _prepareAndExec(dispatchFuncStub);

              dispatchFuncStub
              |> withOneArg(
                   AppStore.UpdateAction(Update([|UpdateStore.Inspector|])),
                 )
              |> getCallCount
              |> expect == dispatchFuncCallCountBeforeChangeTarget
              + 1;
            });
          })
        );
      });

      describe("test change arcballCameraController minDistance", () =>
        test("test change should set into engine", () => {
          MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
          let currentGameObjectArcballCamera =
            GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
          let value = 11.1;

          MainEditorArcballCameraControllerTool.changeMinDistanceAndBlur(
            ~cameraController=currentGameObjectArcballCamera,
            ~value,
            (),
          );

          ArcballCameraEngineService.unsafeGetArcballCameraControllerMinDistance(
            currentGameObjectArcballCamera,
          )
          |> StateLogicService.getEngineStateToGetData
          |. FloatService.truncateFloatValue(5)
          |> expect == value;
        })
      );

      describe("test change arcballCameraController phi", () =>
        test("test change should set into engine", () => {
          MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
          let currentGameObjectArcballCamera =
            GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
          let value = 11.1;

          MainEditorArcballCameraControllerTool.changePhiAndBlur(
            ~cameraController=currentGameObjectArcballCamera,
            ~value,
            (),
          );

          ArcballCameraEngineService.unsafeGetArcballCameraControllerPhi(
            currentGameObjectArcballCamera,
          )
          |> StateLogicService.getEngineStateToGetData
          |. FloatService.truncateFloatValue(5)
          |> expect == value;
        })
      );

      describe("test change arcballCameraController theta", () =>
        test("test change should set into engine", () => {
          MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
          let currentGameObjectArcballCamera =
            GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
          let value = 2.0;

          MainEditorArcballCameraControllerTool.changeThetaAndBlur(
            ~cameraController=currentGameObjectArcballCamera,
            ~value,
            (),
          );

          ArcballCameraEngineService.unsafeGetArcballCameraControllerTheta(
            currentGameObjectArcballCamera,
          )
          |> StateLogicService.getEngineStateToGetData
          |. FloatService.truncateFloatValue(5)
          |> expect == value;
        })
      );

      describe("test change arcballCameraController target", () => {
        describe("test change should set into engine", () => {
          let _test = (changeAndBlurFunc, getTargetTargetFunc) => {
            MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
            let currentGameObjectArcballCamera =
              GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
            let value = 11.1;

            changeAndBlurFunc(
              ~cameraController=currentGameObjectArcballCamera,
              ~value,
              (),
            );

            ArcballCameraEngineService.unsafeGetArcballCameraControllerTarget(
              currentGameObjectArcballCamera,
            )
            |> StateLogicService.getEngineStateToGetData
            |> Vector3Service.truncate(5)
            |> expect == getTargetTargetFunc(value);
            /* ; */
          };

          test("test change target x", () =>
            _test(
              MainEditorArcballCameraControllerTool.changeTargetXAndBlur(
                ~uiState=TestTool.buildEmptyAppState(),
                ~dispatchFunc=TestTool.getDispatch(),
              ),
              value =>
              (value, 0., 0.)
            )
          );
          test("test change target y", () =>
            _test(
              MainEditorArcballCameraControllerTool.changeTargetYAndBlur(
                ~uiState=TestTool.buildEmptyAppState(),
                ~dispatchFunc=TestTool.getDispatch(),
              ),
              value =>
              (0., value, 0.)
            )
          );
          test("test change target z", () =>
            _test(
              MainEditorArcballCameraControllerTool.changeTargetZAndBlur(
                ~uiState=TestTool.buildEmptyAppState(),
                ~dispatchFunc=TestTool.getDispatch(),
              ),
              value =>
              (0., 0., value)
            )
          );
        });

        describe("if blur", () =>
          describe("not refresh inspector", () => {
            let _prepareAndExec = dispatchFuncStub => {
              MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
              MainEditorTransformTool.setLocalEulerAngleData();
              let currentGameObjectArcballCamera =
                GameObjectTool.getCurrentSceneTreeNodeArcballCamera();
              let value = 21.1;

              let dispatchFuncCallCountBeforeChangeTarget =
                dispatchFuncStub |> getCallCount;
              MainEditorArcballCameraControllerTool.changeTargetXAndBlur(
                ~cameraController=currentGameObjectArcballCamera,
                ~value,
                (),
              );

              dispatchFuncCallCountBeforeChangeTarget;
            };

            test(
              "shouldn't remove current scene tree node->local euler angle data",
              () => {
              _prepareAndExec(ReactTool.createDispatchFuncStub(sandbox));

              MainEditorTransformTool.judgeShouldRemoveLocalEulerAngleData()
              |> expect == false;
            });
            test("not refresh inspector", () => {
              let dispatchFuncStub =
                ReactTool.createDispatchFuncStub(sandbox);

              let dispatchFuncCallCountBeforeChangeTarget =
                _prepareAndExec(dispatchFuncStub);

              dispatchFuncStub
              |> withOneArg(
                   AppStore.UpdateAction(Update([|UpdateStore.Inspector|])),
                 )
              |> getCallCount
              |> expect == dispatchFuncCallCountBeforeChangeTarget;
            });
          })
        );
      });

      describe(
        "add shade dom for transformComponent if has arcballCameraController",
        () =>
        test("test snapshot for transform component", () => {
          MainEditorInspectorAddComponentTool.addArcballCameraControllerComponent();
          let currentGameObjectTransform =
            GameObjectTool.getCurrentSceneTreeNodeTransform();
          let component =
            BuildComponentTool.buildMainEditorTransformComponent(
              TestTool.buildEmptyAppState(),
              currentGameObjectTransform,
            );
          component |> ReactTestTool.createSnapshotAndMatch;
        })
      );
    });
  });