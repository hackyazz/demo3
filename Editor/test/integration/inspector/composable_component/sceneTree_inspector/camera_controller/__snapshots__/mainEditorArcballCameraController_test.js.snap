// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditor ArcballCameraController test set currentSceneTreeNode to be camera add shade dom for transformComponent if has arcballCameraController test snapshot for transform component 1`] = `
<article
  className="wonder-inspector-transform"
>
  <article
    className="inspector-item wonder-three-float-input"
  >
    <div
      className="item-header"
      title="position in local coordinate system"
    >
      Position
    </div>
    <div
      className="item-content"
    >
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          X
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </div>
      </article>
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          Y
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </div>
      </article>
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          Z
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="4"
          />
        </div>
      </article>
    </div>
  </article>
  <article
    className="inspector-item wonder-three-float-input"
  >
    <div
      className="item-header"
      title="euler angle in local coordinate system"
    >
      Rotation
    </div>
    <div
      className="item-content"
    >
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          X
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </div>
      </article>
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          Y
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </div>
      </article>
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          Z
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </div>
      </article>
    </div>
  </article>
  <article
    className="inspector-item wonder-three-float-input"
  >
    <div
      className="item-header"
      title="scale in local coordinate system"
    >
      Scale
    </div>
    <div
      className="item-content"
    >
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          X
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="1"
          />
        </div>
      </article>
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          Y
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="1"
          />
        </div>
      </article>
      <article
        className="inspector-item wonder-float-input"
      >
        <div
          className="item-header component-label"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          title=""
        >
          Z
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="1"
          />
        </div>
      </article>
    </div>
  </article>
  <div
    className="transform-shade"
  />
</article>
`;

exports[`MainEditor ArcballCameraController test set currentSceneTreeNode to be camera test ui 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <div
      className="sceneTree-name"
    >
      <article
        className="inspector-item"
      >
        <div
          className="item-header"
          title="gameObject's name"
        >
          <span
            className="component-label"
          >
            Name
          </span>
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="Camera"
          />
        </div>
      </article>
    </div>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
          title="Transform"
        >
          Transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="inspector-item wonder-three-float-input"
        >
          <div
            className="item-header"
            title="position in local coordinate system"
          >
            Position
          </div>
          <div
            className="item-content"
          >
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                X
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Y
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Z
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="4"
                />
              </div>
            </article>
          </div>
        </article>
        <article
          className="inspector-item wonder-three-float-input"
        >
          <div
            className="item-header"
            title="euler angle in local coordinate system"
          >
            Rotation
          </div>
          <div
            className="item-content"
          >
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                X
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Y
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Z
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
          </div>
        </article>
        <article
          className="inspector-item wonder-three-float-input"
        >
          <div
            className="item-header"
            title="scale in local coordinate system"
          >
            Scale
          </div>
          <div
            className="item-content"
          >
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                X
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="1"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Y
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="1"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Z
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="1"
                />
              </div>
            </article>
          </div>
        </article>
        <div
          className="transform-shade"
        />
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
          title="Camera Group set the gameObject's camera related components"
        >
          Camera Group
        </div>
        <div
          className="header-close"
        >
          <img
            onClick={[Function]}
            src="./public/img/close.png"
          />
        </div>
      </div>
      <article
        className="wonder-camera-group"
      >
        <div
          className="inspector-component"
        >
          <div
            className="component-title"
            title="CameraView"
          >
            CameraView
          </div>
          <hr />
          <div
            className="component-content"
          >
            <article
              className="wonder-camera-view"
            >
              <article
                className="inspector-item"
              >
                <div
                  className="item-header"
                  title="type"
                >
                  <span
                    className=""
                  >
                    Type
                  </span>
                </div>
                <div
                  className="item-content"
                >
                  <select
                    onChange={[Function]}
                    value="0"
                  >
                    <option
                      value="0"
                    >
                      basic_cameraView
                    </option>
                  </select>
                </div>
              </article>
              <div
                className="inspector-item"
              >
                <div
                  className="item-header"
                  title="set the camera currently in use. only one camera can be used in the scene. after setting, other cameras will be cancelled."
                >
                  CurrentCamera
                </div>
                <div
                  className="item-content"
                >
                  <input
                    defaultChecked={true}
                    disabled={true}
                    onClick={[Function]}
                    type="checkbox"
                  />
                </div>
              </div>
            </article>
          </div>
        </div>
        <div
          className="inspector-component"
        >
          <div
            className="component-title"
            title="Projection component, responsible for perspective matrix's attributes."
          >
            Projection
          </div>
          <hr />
          <div
            className="component-content"
          >
            <article
              className="wonder-camera-projection"
            >
              <article
                className="inspector-item"
              >
                <div
                  className="item-header"
                  title="perspective matrix's type"
                >
                  <span
                    className=""
                  >
                    Type
                  </span>
                </div>
                <div
                  className="item-content"
                >
                  <select
                    onChange={[Function]}
                    value="0"
                  >
                    <option
                      value="0"
                    >
                      perspective
                    </option>
                  </select>
                </div>
              </article>
              <article
                className="inspector-item wonder-float-input"
              >
                <div
                  className="item-header component-label"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseUp={[Function]}
                  title="up near-plane distance"
                >
                  Near
                </div>
                <div
                  className="item-content"
                >
                  <input
                    className="input-component float-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="text"
                    value="0.01"
                  />
                </div>
              </article>
              <article
                className="inspector-item wonder-float-input"
              >
                <div
                  className="item-header component-label"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseUp={[Function]}
                  title="up far-plane distance"
                >
                  Far
                </div>
                <div
                  className="item-content"
                >
                  <input
                    className="input-component float-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="text"
                    value="50000"
                  />
                </div>
              </article>
              <article
                className="inspector-item wonder-float-input"
              >
                <div
                  className="item-header component-label"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseUp={[Function]}
                  title="define the vertical viewing range of the lens in terms of angle"
                >
                  Fovy
                </div>
                <div
                  className="item-content"
                >
                  <input
                    className="input-component float-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="text"
                    value="60"
                  />
                </div>
              </article>
            </article>
          </div>
        </div>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
          title="arcball camera controller"
        >
          ArcballCameraController
        </div>
        <div
          className="header-close"
        >
          <img
            onClick={[Function]}
            src="./public/img/close.png"
          />
        </div>
      </div>
      <article
        className="wonder-inspector-arcballCameraController"
      >
        <article
          className="inspector-item wonder-float-input"
        >
          <div
            className="item-header component-label"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            title="distance between camera and target"
          >
            Distance
          </div>
          <div
            className="item-content"
          >
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="10"
            />
          </div>
        </article>
        <article
          className="inspector-item wonder-float-input"
        >
          <div
            className="item-header component-label"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            title="min distance between camera and target"
          >
            Min Distance
          </div>
          <div
            className="item-content"
          >
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="0.05"
            />
          </div>
        </article>
        <article
          className="inspector-item wonder-three-float-input"
        >
          <div
            className="item-header"
            title="target"
          >
            Target
          </div>
          <div
            className="item-content"
          >
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                X
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Y
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
            <article
              className="inspector-item wonder-float-input"
            >
              <div
                className="item-header component-label"
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                title=""
              >
                Z
              </div>
              <div
                className="item-content"
              >
                <input
                  className="input-component float-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="0"
                />
              </div>
            </article>
          </div>
        </article>
        <article
          className="inspector-item wonder-float-input"
        >
          <div
            className="item-header component-label"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            title="horizontal radian"
          >
            Phi
          </div>
          <div
            className="item-content"
          >
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="1.5708"
            />
          </div>
        </article>
        <article
          className="inspector-item wonder-float-input"
        >
          <div
            className="item-header component-label"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            title="vertical radian"
          >
            Theta
          </div>
          <div
            className="item-content"
          >
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="1.5708"
            />
          </div>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <div
        className="addable-component-content"
      >
        <div
          className="content-btn"
        >
          <button
            className="addable-btn"
            onClick={[Function]}
          >
            Add Component
          </button>
        </div>
      </div>
    </article>
  </article>
</article>
`;
